### ISSUE TITLE
Usage for SEED Dataset with LOSO and DGCNN Framework

### ISSUE THREAD

**Issue by VerdantE1** [XJTU-EEG/LibEER#2]
Hi,
Thank you for providing this library! I would like to know how to use this library for implementing a Leave-One-Subject-Out (LOSO) cross-validation paradigm using the SEED dataset and the DGCNN (Dynamic Graph Convolutional Neural Network) model.

Could you please provide a clear usage example or step-by-step instructions for:

Preprocessing the SEED dataset for compatibility with the DGCNN framework.
Setting up and running the LOSO cross-validation paradigm.
Training and evaluating the DGCNN model within this setup with the first session only.
If there are specific configurations, hyperparameters, or scripts already available for this purpose, kindly point me to them or provide a minimal working example.

Looking forward to your response! Thanks for your help!

**Comment by XJTU-EEG** [comment]
Hi, VerdantE1!
Thank you for your interest in our work. Based on your description, you need to use DGCNN to complete the LOSO task on the SEED dataset.

We have described in the Readme that the main configurations of our algorithm library are primarily set through the `Setting` class. If you would like to learn more about the specific parameter configurations, you can refer to the [Guide about setting parameters](docs/SettingParameters.md)，and the pre-configured settings we have provided:[Preset Setting in LibEER](docs/PresetSetting.md)。

Based on your requirements, I have prepared two feasible solutions for you. Both of the following solutions are ready to run directly, and detailed comments have been provided for your reference.
## First Solution
```python
from models.Models import Model
from config.setting import  Setting
from data_utils.load_data import get_data
from data_utils.split import merge_to_part, index_to_data, get_split_index
from utils.args import get_args_parser
from utils.store import make_output_dir
from utils.utils import result_log, setup_seed
from Trainer.training import train
from models.DGCNN import NewSparseL2Regularization
import torch
import torch.optim as optim
import torch.nn as nn


def main(args):
    setting = Setting(dataset='seed_de_lds', # Select the dataset -> For the SEED dataset, we recommend using the features provided by the SEED dataset.
                                             # Among them, the DE_LDS features achieve the best results in experiments, using the DE_LDS features provided by the SEED dataset.
                                             # For the choice of dataset, refer to the SettingParameters in the docs folder.
                      dataset_path='yourDatasetPath/SEED/', # Specify the path to the corresponding dataset.
                      sessions=[1], # Specify using only the first session of the SEED dataset.
                      sample_length=1, # Specify the sample length. Since the input of DGCNN does not have time steps, this is set to 1, meaning the input sample length is 1.
                      stride=1, # Specify the step size for sampling. Sliding window sampling is used, and this specifies the window's movement size.
                                # If it is the same as sample_length, it means there is no overlap between the samples.
                      seed=2024, # Set the random seed for the experiment.
                      experiment_mode="subject-independent", # Specify the experiment task, i.e., the subject-independent task.
                      split_type='leave-one-out', # Specify the data splitting method. Here, it is set to leave-one-out.
                       )
    setup_seed(2024) # Set the random seed for the experiment to ensure reproducibility.
    data, label, channels, feature_dim, num_classes = get_data(setting) # Get the corresponding data and information based on the setting class.
    # The organization of data and label is [session(1), subject(15), trial(15), sample(XXX)].
    data, label = merge_to_part(data, label, setting) # Merge the data based on the experiment task specified in the setting class.
    # After the merge_to_part() function and the specified subject-independent method, the organization of data and label will be [[subject(15), sample(xxx)]].
    device = torch.device(args.device) # Set the device based on the args command-line parameters.
    best_metrics = [] # Prepare to record the experimental results.
    for rridx, (data_i, label_i) in enumerate(zip(data, label), 1): # This loop will only execute once; it will be enabled only under the subject-dependent task.
        tts = get_split_index(data_i, label_i, setting) # Get the task indexes for the experiment based on the setting class. The leave-one-out splitting method was chosen.
        # Here, in tts:
        # train indexes: [1, 2, 3, ..., 14], test indexes: [0]
        # train indexes: [0, 2, 3, ..., 14], test indexes: [1]
        # ...
        # Each individual's data will be used as the test set for one experiment.
        for ridx, (train_indexes, test_indexes, val_indexes) in enumerate(zip(tts['train'], tts['test'], tts['val']), 1):
            setup_seed(args.seed) # Set the random seed again to ensure reproducibility.
            if val_indexes[0] == -1:
                print(f"train indexes:{train_indexes}, test indexes:{test_indexes}")
            else:
                print(f"train indexes:{train_indexes}, val indexes:{val_indexes}, test indexes:{test_indexes}")

            # Retrieve the corresponding data based on the indexes. train_data contains data from 14 individuals, and test_data contains data from one individual.
            train_data, train_label, val_data, val_label, test_data, test_label = \
                index_to_data(data_i, label_i, train_indexes, test_indexes, val_indexes)
            # model to train
            if len(val_data) == 0:
                val_data = test_data
                val_label = test_label
            # Choose a model. Alternatively, you can use the method below to import the DGCNN model:
            # model = DGCNN(channels, feature_dim, num_classes)
            # You can configure the model parameters in model_param/DGCNN.yaml
            model = Model['DGCNN'](channels, feature_dim, num_classes)
            # Prepare the corresponding dataloader.
            dataset_train = torch.utils.data.TensorDataset(torch.Tensor(train_data), torch.Tensor(train_label))
            dataset_val = torch.utils.data.TensorDataset(torch.Tensor(val_data), torch.Tensor(val_label))
            dataset_test = torch.utils.data.TensorDataset(torch.Tensor(test_data), torch.Tensor(test_label))
            # Select an appropriate optimizer.
            optimizer = optim.AdamW(model.parameters(), lr=args.lr, weight_decay=1e-4, eps=1e-4)
            # Select appropriate loss functions. The first is a classification loss function, and the second is the L2 regularization loss in DGCNN.
            criterion = nn.CrossEntropyLoss()
            loss_func = NewSparseL2Regularization(0.01).to(device)
            # Specify the output_dir, mainly for saving intermediate results during model training. It is set based on args but may show errors currently.
            output_dir = make_output_dir(args, "DGCNN")
            # Call the training function to train. Batch size, epochs, etc., can be set via command-line parameters, or manually if desired.
            round_metric = train(model=model, dataset_train=dataset_train, dataset_val=dataset_val, dataset_test=dataset_test, device=device,
                                 output_dir=output_dir, metrics=args.metrics, metric_choose=args.metric_choose, optimizer=optimizer,
                                 batch_size=args.batch_size, epochs=args.epochs, criterion=criterion, loss_func=loss_func, loss_param=model)
            best_metrics.append(round_metric)
    # best metrics: every round metrics dict
    result_log(args, best_metrics)

if __name__ == '__main__':
    args = get_args_parser()
    args = args.parse_args()
    main(args)
```

## Second Solution

We have also prepared a `PresetSetting` class, which includes the LOSO task on the SEED dataset, using only the data from the first session.

```python
# This only shows the configuration for the setting class
# -dataset seed_de_lds -dataset_path 'yourDatasetPath/SEED/'
setting = preset_setting["seed_sub_independent_leave_one_out_setting"](args)
```

When running the script, include the following parameters:

```cmd
python YourDataName.py -dataset seed_de_lds -dataset_path 'yourDatasetPath/SEED/'
```
